For a given subscription, last_seq is lower than last_sent

Hello,

I've observed weird subscription behavior, where `last_sent` is far above the `last_seq` value. For instance:
```
curl -X GET "http://localhost:8222/streaming/channelsz?channel=sub_name&subs=1"                                  
{                                                                                                                                              
  "name": "sub_name",                                                                                                                       
  "msgs": 9,                                                
  "bytes": 14607,                                                                                                                                                               
  "first_seq": 1,                                                        
  "last_seq": 9,                                                                                                                                                                
  "subscriptions": [                                           
    {                                                                                                                                                                           
      "client_id": "basket",                                 
      "inbox": "ef8a68605472c1df7b88636838",                                                                                                    
      "ack_inbox": "_INBOX.mp6XYgPKFRe9PK9XiQW0gv",
      "queue_name": "basket:basket",               
      "is_durable": true,           
      "is_offline": false, 
      "max_inflight": 1024,
      "ack_wait": 30,                                                                                                                                                           
      "last_sent": 57,                                                  
      "pending_count": 0,                                                                                                                                                      
      "is_stalled": false                                                                                                                      
    }                                                                                                                                                                           
  ]  
```

The result of that is all messages are being ignored until `last_seq > last_sent`. Is it expected behavior? If so, how to avoid it? 

Thanks!

@kozlovic , sorry for late reply.

I've digged into it today and failed while trying to compile(I am not golang guy). Here is what I've tried:
```
go build
nats-streaming-server.go:28:2: cannot find package "github.com/lib/pq" in any of:
	/usr/lib/go-1.10/src/github.com/lib/pq (from $GOROOT)
       ...
```

so I've found `vendor` dir with these missing deps, but I don't know how to use it , sorry for that. I've also tried `scripts/cross_compile.sh` , but it failed `env: ‘gox’: No such file or directory`. I'd be glad if you could provide me a few steps how to deal with it. Thanks!

--- a/stores/sqlstore.go
+++ b/stores/sqlstore.go
@@ -817,7 +817,8 @@ func (s *SQLStore) Recover() (*RecoveredState, error) {
 		var (
 			channelID int64
 			name      string
-			maxseq    uint64
+			maxseq    uint64 // We get that from the Channels table.
+			mmseq     uint64 // This is the max seq found in the Messages table for given channel.
 		)
 		if err := channelRows.Scan(&channelID, &name, &maxseq); err != nil {
 			return nil, err
@@ -826,6 +827,16 @@ func (s *SQLStore) Recover() (*RecoveredState, error) {
 		channelLimits := s.genericStore.getChannelLimits(name)
 
 		msgStore := s.newSQLMsgStore(name, channelID, &channelLimits.MsgStoreLimits)
+
+		// We need to get the last seq from messages table before possibly expiring messages.
+		r = s.preparedStmts[sqlGetLastSeq].QueryRow(channelID)
+		if err := r.Scan(&mmseq); err != nil {
+			return nil, sqlStmtError(sqlGetLastSeq, err)
+		}
+		// If it is more than the one that was updated in the Channel row, then use this one.
+		if mmseq > maxseq {
+			maxseq = mmseq
+		}
 
 		if err := s.applyLimitsOnRecovery(msgStore); err != nil {
 			return nil, err
@@ -846,10 +857,9 @@ func (s *SQLStore) Recover() (*RecoveredState, error) {
 		msgStore.last = last
 		msgStore.totalCount = totalCount
 		msgStore.totalBytes = totalBytes
-		// If all messages have expired, the above should all be 0, however,
-		// the Channel table may contain a maxseq that we should use as starting
-		// point.
-		if msgStore.last == 0 {
+		// Since messages may have been removed due to limits, update first/last
+		// based on known max sequence.
+		if maxseq > msgStore.last {
 			msgStore.first = maxseq + 1
 			msgStore.last = maxseq
 		}

--- a/stores/sqlstore_test.go
+++ b/stores/sqlstore_test.go
@@ -1716,3 +1716,38 @@ func TestSQLSubStoreFlush(t *testing.T) {
 		t.Fatalf("Expected last sent to be %v, got %v", lastSent, rs.Sub.LastSent)
 	}
 }
+
+func TestSQLRecoverLastSeqAfterMessagesExpired(t *testing.T) {
+	if !doSQL {
+		t.SkipNow()
+	}
+	cleanupSQLDatastore(t)
+	defer cleanupSQLDatastore(t)
+
+	s := createDefaultSQLStore(t)
+	defer s.Close()
+
+	cs := storeCreateChannel(t, s, "foo")
+	payload := make([]byte, 100)
+	storeMsg(t, cs, "foo", 1, payload)
+	storeMsg(t, cs, "foo", 2, payload)
+	storeMsg(t, cs, "foo", 3, payload)
+	storeMsg(t, cs, "foo", 4, payload)
+	storeMsg(t, cs, "foo", 5, payload)
+	s.Close()
+
+	// Sleep for more than the maxAge we will then set
+	time.Sleep(700 * time.Millisecond)
+
+	// Restart store with a maxAge
+	limits := testDefaultStoreLimits
+	limits.MaxAge = 250 * time.Millisecond
+	s, state := openDefaultSQLStoreWithLimits(t, &limits)
+	defer s.Close()
+	cs = state.Channels["foo"].Channel
+	first, last := msgStoreFirstAndLastSequence(t, cs.Msgs)
+	if first != 6 && last != 5 {
+		t.Fatalf("Should be left with 6..5, got %v..%v", first, last)
+	}
+	s.Close()
+}

